com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v1/ApplicationsCloudModelBuilder.java

```diff
@@ -99,10 +100,21 @@ protected ApplicationEnvironmentCloudModelBuilder createApplicationEnvironmentCl
     }
 
     public List<CloudApplicationExtended> build(Set<String> mtaModulesInArchive, Set<String> allMtaModules, Set<String> deployedModules) {
-        List<CloudApplicationExtended> apps = new ArrayList<>();
         SortedSet<String> unresolvedMtaModules = new TreeSet<>(allMtaModules);
         initializeModulesDependecyTypes(deploymentDescriptor);
-        for (Module module : handler.getSortedModules(deploymentDescriptor, SupportedParameters.DEPENDENCY_TYPE, DEPENDECY_TYPE_HARD)) {
+        List<CloudApplicationExtended> apps = resolveModules(mtaModulesInArchive, allMtaModules, deployedModules, unresolvedMtaModules);
+        unresolvedMtaModules.removeAll(deployedModules);
+        if (!unresolvedMtaModules.isEmpty()) {
+            throw new ContentException(Messages.UNRESOLVED_MTA_MODULES, unresolvedMtaModules);
+        }
+        return apps;
+    }
+
+    private List<CloudApplicationExtended> resolveModules(Set<String> mtaModulesInArchive, Set<String> allMtaModules,
+        Set<String> deployedModules, SortedSet<String> unresolvedMtaModules) {
+        List<CloudApplicationExtended> apps = new ArrayList<>();
+        for (Module module : handler.getModulesForDeployment(deploymentDescriptor, SupportedParameters.PARALLEL_DEPLOYMENTS,
+            SupportedParameters.DEPENDENCY_TYPE, DEPENDENCY_TYPE_HARD)) {
             if (!mtaModulesInArchive.contains(module.getName()) || module.getType() == null) {

```

Comment:
I am hurt when I see this logic below. As you are changing it a bit, is there any way to refactor it a bit?